// surrealdb.js/script/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.surrealql = exports.surql = exports.PreparedQuery = exports.ExperimentalSurrealHTTP = exports.SurrealWebSocket = exports.Surreal = void 0;
const websocket_js_1 = require("./strategies/websocket.js");
Object.defineProperty(exports, "Surreal", { enumerable: true, get: function () { return websocket_js_1.WebSocketStrategy; } });
Object.defineProperty(exports, "SurrealWebSocket", { enumerable: true, get: function () { return websocket_js_1.WebSocketStrategy; } });
const http_js_1 = require("./strategies/http.js");
Object.defineProperty(exports, "ExperimentalSurrealHTTP", { enumerable: true, get: function () { return http_js_1.HTTPStrategy; } });
exports.default = websocket_js_1.WebSocketStrategy;
var PreparedQuery_js_1 = require("./library/PreparedQuery.js");
Object.defineProperty(exports, "PreparedQuery", { enumerable: true, get: function () { return PreparedQuery_js_1.PreparedQuery; } });
var tagged_template_js_1 = require("./library/tagged-template.js");
Object.defineProperty(exports, "surql", { enumerable: true, get: function () { return tagged_template_js_1.surql; } });
Object.defineProperty(exports, "surrealql", { enumerable: true, get: function () { return tagged_template_js_1.surrealql; } });
