// surrealdb.js/esm/library/SurrealHTTP.js
import { NoConnectionDetails } from "../errors.js";
import { processUrl } from "./processUrl.js";
export class SurrealHTTP {
    constructor(url, { fetch: f } = {}) {
        Object.defineProperty(this, "url", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "authorization", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "fetch", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "namespace", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "database", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fetch = f ?? fetch;
        this.url = processUrl(url, {
            ws: "http",
            wss: "https",
        });
    }
    ready() {
        return !!(this.url && this.namespace && this.database);
    }
    setTokenAuth(token) {
        this.authorization = `Bearer ${token}`;
    }
    clearAuth() {
        this.authorization = undefined;
    }
    async request(path, options) {
        path = path.startsWith("/") ? path.slice(1) : path;
        if (!this.ready())
            throw new NoConnectionDetails();
        return (await this.fetch(`${this.url}/${path}${options?.searchParams ? `?${options?.searchParams}` : ""}`, {
            method: options?.method ?? "POST",
            headers: {
                "Content-Type": options?.plainBody
                    ? "text/plain"
                    : "application/json",
                Accept: "application/json",
                NS: this.namespace,
                DB: this.database,
                ...(this.authorization
                    ? { Authorization: this.authorization }
                    : {}),
            },
            body: typeof options?.body == "string"
                ? options?.body
                : JSON.stringify(options?.body),
        })).json();
    }
}
